File: BHAVRP_Java/BHAVRPv2.0/BHAVRP2.0/src/cujae/inf/ic/om/assignment/clustering/hierarchical/UPGMC.java
package cujae.inf.ic.om.assignment.clustering.hierarchical;

import java.util.ArrayList;

import cujae.inf.ic.om.factory.DistanceType;

import cujae.inf.ic.om.problem.input.Customer;
import cujae.inf.ic.om.problem.input.Depot;
import cujae.inf.ic.om.problem.input.Location;
import cujae.inf.ic.om.problem.input.Problem;

import cujae.inf.ic.om.problem.output.Cluster;
import cujae.inf.ic.om.problem.output.Solution;
import cujae.inf.ic.om.service.OSRMService;

import cujae.inf.ic.om.matrix.NumericMatrix;
import cujae.inf.ic.om.matrix.RowCol;

public class UPGMC extends Hierarchical {

	public static DistanceType distanceType = DistanceType.Manhattan;
	private static Solution solution = new Solution();
	
	private ArrayList<Integer> listIDElements;
	private ArrayList<Cluster> listClusters;
	private ArrayList<Customer> listCustomersToAssign;
	private ArrayList<Depot> listDepots;
	private NumericMatrix costMatrix;
	
	public UPGMC() {
		super();
	}

	@Override
	public Solution toClustering() {
		initialize();
		assign();
		return finish();
	}
			
	@Override
	public void initialize() {
		listIDElements = Problem.getProblem().getListIDElements();
		listClusters = initializeClusters(listIDElements);
		listCustomersToAssign = new ArrayList<Customer>(Problem.getProblem().getCustomers());
		listDepots = new ArrayList<Depot>(Problem.getProblem().getDepots());
	}
	
	@Override
	public void assign() {
		int totalDepots = listDepots.size();
		
		int currentDepots =  -1;
		int currentCustomers = -1;

		int posCol = -1;
		int posRow = -1;

		int idCustomerOne = -1;
		int posCustomerOne = -1;
		int idCustomerTwo = -1;
		int posCustomerTwo = -1;

		int posClusterOne = -1;
		double requestClusterOne = 0.0;
		int posClusterTwo = -1;
		double requestClusterTwo = 0.0;

		int idDepot = -1;
		int posDepot = -1;		
		int posDepotMatrix = -1;
		double capacityDepot = 0.0;

		int posCluster = -1;
		double requestCluster = 0.0;
		double totalRequest = 0.0;

		int idDepotWithMU = -1;
		double capacityDepotWithMU = 0.0;
		RowCol rcBestAll = new RowCol();
		Location newLocation = new Location();
		
		boolean change = true;
		
		while(!(listCustomersToAssign.isEmpty()) && (totalDepots > 0))
		{
			currentDepots = listDepots.size();
			currentCustomers = listCustomersToAssign.size();

			System.out.println("Current Depots: " + currentDepots);
			System.out.println("Current Customers: " + currentCustomers);
			
			if(change)
			{
				try {
					costMatrix = initializeCostMatrix(listCustomersToAssign, listDepots, distanceType);
				} catch (IllegalArgumentException | SecurityException e) {
					e.printStackTrace();
				}
			}
			
			rcBestAll = costMatrix.indexLowerValue(0, 0, (currentCustomers + currentDepots - 1), (currentCustomers - 1));
			
			System.out.println("rcBestAll " + costMatrix.getItem(rcBestAll.getRow(), rcBestAll.getCol()));
			System.out.println("bestAllRow" + rcBestAll.getRow());
			System.out.println("bestAllCol" + rcBestAll.getCol());
			
			posCol = rcBestAll.getCol();
			posRow = rcBestAll.getRow();

			if((posCol < currentCustomers) && (posRow < currentCustomers)) 
			{ 
				posCustomerOne = posCol;
				idCustomerOne = listCustomersToAssign.get(posCustomerOne).getIDCustomer();

				System.out.println("--------------------------------------");
				System.out.println("ID Customer One: " + idCustomerOne);
				
				posCustomerTwo = posRow;
				idCustomerTwo = listCustomersToAssign.get(posCustomerTwo).getIDCustomer();				

				System.out.println("idCustomerTwo" + idCustomerTwo);
				
				posClusterOne = findCluster(idCustomerOne, listClusters);	
				posClusterTwo = findCluster(idCustomerTwo, listClusters);
				
				System.out.println("--------------------------------------");
				System.out.println("Position Cluster One: " + posClusterOne);
				System.out.println("Position Cluster Two: " + posClusterTwo);
				
				if((posClusterOne != -1) && (posClusterTwo != -1)) 
				{
					requestClusterOne = listClusters.get(posClusterOne).getRequestCluster();
					requestClusterTwo = listClusters.get(posClusterTwo).getRequestCluster();
					totalRequest = requestClusterOne + requestClusterTwo;
					
					System.out.println("--------------------------------------");
					System.out.println("Request Cluster One: " + requestClusterOne);
					System.out.println("Request Cluster Two: " + requestClusterTwo);
					System.out.println("Total Request: " + totalRequest);
					
					idDepotWithMU = getIDClusterWithMU(listDepots, listClusters);
					posDepot = Problem.getProblem().findPosDepot(listDepots, idDepotWithMU);
					capacityDepotWithMU = Problem.getProblem().getTotalCapacityByDepot(listDepots.get(posDepot));
					posCluster = findCluster(idDepotWithMU, listClusters);
					requestCluster = listClusters.get(posCluster).getRequestCluster();

					System.out.println("--------------------------------------");
					System.out.println("ID Depot With MU: " + idDepotWithMU);
					System.out.println("Position Depot: " + posDepot);
					System.out.println("Capacity Depot With MU: " + capacityDepotWithMU);
					System.out.println("Position Cluster: " + posCluster);
					System.out.println("Request Cluster: " + requestCluster);
					
					if(capacityDepotWithMU >= (requestCluster + totalRequest)) 
					{
						listClusters.get(posClusterOne).setRequestCluster(totalRequest);

						for (int i = 0; i < listClusters.get(posClusterTwo).getItemsOfCluster().size(); i++)
							listClusters.get(posClusterOne).getItemsOfCluster().add(listClusters.get(posClusterTwo).getItemsOfCluster().get(i));			

						newLocation = recalculateCentroid(listClusters.get(posClusterOne));
						
						System.out.println("--------------------------------------");
						System.out.println("New Location:");
						System.out.println("Axis X: " + newLocation.getAxisX());
						System.out.println("Axis Y: " + newLocation.getAxisY());
						
						posCustomerOne = Problem.getProblem().findPosCustomer(listCustomersToAssign, idCustomerOne);
						listCustomersToAssign.get(posCustomerOne).setLocationCustomer(newLocation);
						
						System.out.println("--------------------------------------");
						System.out.println("New Location:");
						System.out.println("Axis X: " + newLocation.getAxisX());
						System.out.println("Axis Y:" + newLocation.getAxisY());
						System.out.println("Position Customer One: " + posCustomerOne);
						
						listClusters.remove(posClusterTwo);
						posCustomerTwo = Problem.getProblem().findPosCustomer(listCustomersToAssign, idCustomerTwo);
						listCustomersToAssign.remove(posCustomerTwo);
						
						System.out.println("--------------------------------------");
						System.out.println("List Clusters: " + listClusters.size());
						System.out.println("Position Customer Two: " + posCustomerTwo);
						System.out.println("List Customers To Assign: " + listCustomersToAssign.size());
						
						change = true;
					}
					else
					{
						costMatrix.setItem(rcBestAll.getRow(), rcBestAll.getCol(), Double.POSITIVE_INFINITY);
						change = false;
					}
				}	
			}
			else
			{				
				if(((posCol < currentCustomers) && (posRow >= currentCustomers)) || ((posRow < currentCustomers) && (posCol >= currentCustomers)))
				{
					if(posCol < currentCustomers)
					{
						posCustomerOne = rcBestAll.getCol();
						posDepotMatrix = rcBestAll.getRow();
					}
					else
					{
						posCustomerOne = rcBestAll.getRow();
						posDepotMatrix = rcBestAll.getCol();
					}
					
					System.out.println("--------------------------------------");
					System.out.println("Position Customer One: " + posCustomerOne);
					System.out.println("Position Depot Matrix: " + posDepotMatrix);

					idCustomerOne = listCustomersToAssign.get(posCustomerOne).getIDCustomer();			
					posClusterOne = findCluster(idCustomerOne, listClusters);
					
					System.out.println("--------------------------------------");
					System.out.println("ID Customer One: " + idCustomerOne);
					System.out.println("Position Cluster One: " + posClusterOne);

					posDepot = (posDepotMatrix - currentCustomers); 
					idDepot = listDepots.get(posDepot).getIDDepot();
					capacityDepot = Problem.getProblem().getTotalCapacityByDepot(listDepots.get(posDepot));
					posCluster = findCluster(idDepot, listClusters);

					System.out.println("--------------------------------------");
					System.out.println("Position Depot: " + posDepot);
					System.out.println("ID Depot: " + idDepot);
					System.out.println("Capacity Depot: " + capacityDepot);
					System.out.println("Position Cluster: " + posCluster);
					
					if((posClusterOne != -1) && (posCluster != -1))
					{
						requestClusterOne = listClusters.get(posClusterOne).getRequestCluster();
						requestCluster = listClusters.get(posCluster).getRequestCluster();

						System.out.println("--------------------------------------");
						System.out.println("Request Cluster One: " + requestClusterOne);
						System.out.println("Request Cluster: " + requestCluster);
						
						if(capacityDepot >= (requestCluster + requestClusterOne)) 
						{
							requestCluster += requestClusterOne;
							listClusters.get(posCluster).setRequestCluster(requestCluster);

							for (int i = 0; i < listClusters.get(posClusterOne).getItemsOfCluster().size(); i++)
								listClusters.get(posCluster).getItemsOfCluster().add(listClusters.get(posClusterOne).getItemsOfCluster().get(i));
					
							listCustomersToAssign.remove(Problem.getProblem().findPosCustomer(listCustomersToAssign, idCustomerOne));
							
							listClusters.remove(posClusterOne);
							change = true;
						}
						else
						{
							costMatrix.setItem(rcBestAll.getRow(), rcBestAll.getCol(), Double.POSITIVE_INFINITY);
							change = false;		
						}
						
						//PENDIENTE ENTRADA CON LISTCUSTOMER VACIA EN EL ISFULLDEPOT TRADICIONAL
						if(isFullDepot(listClusters, requestCluster, capacityDepot, listCustomersToAssign.size()))
						{
							posCluster = findCluster(idDepot, listClusters);
							
							System.out.println("--------------------------------------");
							System.out.println("Position Cluster: " + posCluster);
							
							if(!(listClusters.get(posCluster).getItemsOfCluster().isEmpty()))
								solution.getClusters().add(listClusters.remove(posCluster));
							else
								listClusters.remove(posCluster);

							listDepots.remove(posDepot);
							totalDepots--;
							change = true;
						}
					}
				}
			}
		}
	}
	
	@Override
	public Solution finish() {	
		finish(listClusters, solution);
		
		if(!listClusters.isEmpty())
			for(int k = 0; k < listClusters.size(); k++)
				if(!(listClusters.get(k).getItemsOfCluster().isEmpty()))
					solution.getClusters().add(listClusters.get(k));

		OSRMService.clearDistanceCache();
		
		return solution;
	}
	
	/*Método encargado de obtener el id del deposito con mayor capacidad de la lista*/
	private int getIDClusterWithMU(ArrayList<Depot> depots, ArrayList<Cluster> clusters) {
		int idDepotMU = depots.get(0).getIDDepot();
		int posCluster = findCluster(idDepotMU, clusters);
		double requestCluster = clusters.get(posCluster).getRequestCluster();
		double maxCapacityDepot = Problem.getProblem().getTotalCapacityByDepot(depots.get(0)); 
		maxCapacityDepot -= requestCluster;
		int totalDepots = depots.size();

		double currentCapacityDepot; 

		for(int i = 1; i < totalDepots; i++)
		{
			posCluster = findCluster(depots.get(i).getIDDepot(), clusters);
			requestCluster = clusters.get(posCluster).getRequestCluster();
			currentCapacityDepot = Problem.getProblem().getTotalCapacityByDepot(depots.get(i));
			currentCapacityDepot -= requestCluster;
			
			if(maxCapacityDepot < currentCapacityDepot)
			{
				maxCapacityDepot = currentCapacityDepot;
				idDepotMU = depots.get(i).getIDDepot(); 
			}
		}
		return idDepotMU;
	}
	
	private void finish(ArrayList<Cluster> clusters, Solution solution){
		int posElement = -1;

		for(int i = 0; i < clusters.size(); i++)
		{
			System.out.println();
			System.out.println("--------------------------------------");
			System.out.println("ID Cluster: " + clusters.get(i).getIDCluster());
			System.out.println("Position Cluster: " + clusters.get(i).getItemsOfCluster().toString());
			System.out.println("Request Cluster: " + clusters.get(i).getRequestCluster());
			
			posElement = Problem.getProblem().findPosCustomer(Problem.getProblem().getCustomers(), clusters.get(i).getIDCluster());

			if(posElement != -1)
			{
				for(int j = 0; j < clusters.get(i).getItemsOfCluster().size(); j++)
					solution.getUnassignedItems().add(clusters.get(i).getItemsOfCluster().get(j).intValue());

				clusters.remove(i);
			}
		}
	}
}

File: BHAVRP_Python/cujae/inf/citi/om/assignment/clustering/hierarchical/upgmc.py
import numpy as np
from typing import List
from .hierarchical import Hierarchical
from ....service.osrm_service import OSRMService
from ....problem.input.problem import Problem
from ....problem.input.customer import Customer
from ....problem.input.depot import Depot
from ....problem.input.location import Location
from ....problem.output.solution import Solution
from ....problem.output.cluster import Cluster

class UPGMC(Hierarchical):
    
    def __init__(self):
        super().__init__()
        self.solution = Solution()
        self.list_id_elements: List[int]
        self.list_clusters: List[Cluster]
        self.list_customers_to_assign: List[Customer]
        self.list_depots: List[Depot]
            
    def to_clustering(self) -> Solution:
        self.initialize()
        self.assign()
        return self.finish()
        
    def initialize(self):
        self.list_id_elements = list(Problem.get_problem().get_list_id_elements())
        self.list_clusters = self.initialize_clusters()
        self.list_customers_to_assign = list(Problem.get_problem().get_customers())
        self.list_depots = list(Problem.get_problem().get_depots())

    def assign(self):
        change: bool = True
        total_depots = len(self.list_depots)

        while any(self.list_customers_to_assign) and len(self.list_depots) > 0:
            current_customers = len(self.list_customers_to_assign)
            current_depots = len(self.list_depots)
            
            print(f"Current Customers: {current_customers}")
            print(f"Current Depots: {current_depots}")
            
            if change:
                cost_matrix: np.ndarray = self.initialize_cost_matrix(self.list_customers_to_assign, self.list_depots, self.distance_type)
        
            min_value = np.min(cost_matrix) 
            row_best, col_best = np.where(cost_matrix == min_value)
            row_best, col_best = row_best[0], col_best[0]
            
            print(f"Minimun Value: {min_value}")
            print(f"Row Best: {row_best}")
            print(f"Col Best: {col_best}")
            
            if col_best < current_customers and (row_best + current_customers) < current_customers:
                id_customer_one = self.list_customers_to_assign[col_best].get_id_customer()
                id_customer_two = self.list_customers_to_assign[row_best].get_id_customer()

                print(f"ID Customer One: {id_customer_one}")
                print(f"ID Customer Two: {id_customer_two}")
                
                pos_cluster_one = self.find_cluster(id_customer_one, self.list_clusters)
                pos_cluster_two = self.find_cluster(id_customer_two, self.list_clusters)
                
                print(f"Position Cluster One: {pos_cluster_one}")
                print(f"Position Cluster Two: {pos_cluster_two}")
                
                if pos_cluster_one != -1 and pos_cluster_two != -1:
                    request_cluster_one = self.list_clusters[pos_cluster_one].get_request_cluster()
                    request_cluster_two = self.list_clusters[pos_cluster_two].get_request_cluster()
                    total_request = request_cluster_one + request_cluster_two
                    
                    print(f"Request Cluster One: {request_cluster_one}")
                    print(f"Request Cluster Two: {request_cluster_two}")
                    print(f"Total Request: {total_request}")

                    id_depot_with_mu = self.get_id_cluster_with_mu(self.list_depots, self.list_clusters)
                    pos_depot = Problem.get_problem().find_pos_depot(self.list_depots, id_depot_with_mu)
                    capacity_depot_with_mu = Problem.get_problem().get_total_capacity_by_depot(self.list_depots[pos_depot])
                    
                    pos_cluster = self.find_cluster(id_depot_with_mu, self.list_clusters)
                    request_cluster = self.list_clusters[pos_cluster].get_request_cluster()
                    
                    print(f"ID Depot with MU: {id_depot_with_mu}")
                    print(f"Position Depot: {pos_depot}")
                    print(f"Capacity Depot with MU: {capacity_depot_with_mu}")
                    print(f"Position Cluster: {pos_cluster}")
                    print(f"Request Cluster: {request_cluster}")

                    if capacity_depot_with_mu >= (request_cluster + total_request):

                        self.list_clusters[pos_cluster_one].set_request_cluster(total_request)
                        self.list_clusters[pos_cluster_one].get_items_of_cluster().extend(self.list_clusters[pos_cluster_two].get_items_of_cluster())
                        
                        new_location: Location = self.recalculate_centroid(self.list_clusters[pos_cluster_one])
                        
                        print(f"New Location X: {new_location.get_axis_x()}")
                        print(f"New Location Y: {new_location.get_axis_y()}")

                        pos_customer_one = Problem.get_problem().find_pos_customer(self.list_customers_to_assign, id_customer_one)
                        self.list_customers_to_assign[pos_customer_one].set_location_customer(new_location)
                        
                        print(f"New Location X: {new_location.get_axis_x()}")
                        print(f"New Location Y: {new_location.get_axis_y()}")
                        print(f"Position Customer One: {pos_customer_one}")                        

                        self.list_clusters.pop(pos_cluster_two)
                        pos_customer_two = Problem.get_problem().find_pos_customer(self.list_customers_to_assign, id_customer_two)
                        self.list_customers_to_assign.pop(pos_customer_two)
                        
                        print(f"List Clusters: {len(self.list_clusters)}")
                        print(f"Position Customer Two: {pos_customer_two}")
                        print(f"List Customers to Assign: {len(self.list_customers_to_assign)}")
                        
                        change = True
                    else:
                        cost_matrix[row_best, col_best] = float("inf")
                        change = False

            elif ((col_best < current_customers <= (row_best + current_customers)) or ((row_best + current_customers) < current_customers <= col_best)):
                if col_best < current_customers:
                    pos_customer_one, pos_depot_matrix = col_best, row_best
                else:
                    pos_customer_one, pos_depot_matrix = row_best, col_best
                    
                print("--------------------------------------")
                print(f"Position Customer One: {pos_customer_one}")
                print(f"Position Depot Matrix: {pos_depot_matrix}")

                id_customer_one = self.list_customers_to_assign[pos_customer_one].get_id_customer()
                pos_cluster_one = self.find_cluster(id_customer_one, self.list_clusters)
                
                print("--------------------------------------")
                print(f"ID Customer One: {id_customer_one}")
                print(f"Position Cluster One: {pos_cluster_one}")

                id_depot = self.list_depots[pos_depot_matrix].get_id_depot()
                capacity_depot = Problem.get_problem().get_total_capacity_by_depot(self.list_depots[pos_depot_matrix])
                pos_cluster = self.find_cluster(id_depot, self.list_clusters)
                
                print("--------------------------------------")
                print(f"Position Depot: {pos_depot_matrix}")
                print(f"ID Depot: {id_depot}")
                print(f"Capacity Depot: {capacity_depot}")
                print(f"Position Cluster: {pos_cluster}")

                if pos_cluster_one != -1 and pos_cluster != -1:
                    request_cluster_one = self.list_clusters[pos_cluster_one].get_request_cluster()
                    request_cluster = self.list_clusters[pos_cluster].get_request_cluster()
                    
                    print("--------------------------------------")
                    print(f"Request Cluster One: {request_cluster_one}")
                    print(f"Request Cluster: {request_cluster}")
                    print("--------------------------------------")

                    if capacity_depot >= request_cluster + request_cluster_one:
                        capacity_depot -= request_cluster + request_cluster_one
                        self.list_clusters[pos_cluster].set_request_cluster(request_cluster + request_cluster_one)
                        self.list_clusters[pos_cluster].get_items_of_cluster().extend(self.list_clusters[pos_cluster_one].get_items_of_cluster())
                        self.list_customers_to_assign.pop(Problem.get_problem().find_pos_customer(self.list_customers_to_assign, id_customer_one))
                        self.list_clusters.pop(pos_cluster_one)
                        change = True
                    else:
                        cost_matrix[row_best, col_best] = float("inf")
                        change = False

                # Verificar si el depósito está lleno
                if self.is_full_depot(self.list_clusters, len(self.list_customers_to_assign), request_cluster, capacity_depot):
                    pos_cluster = self.find_cluster(id_depot, self.list_clusters)
                    print(f"Position Cluster: {pos_cluster}")
                    
                    if self.list_clusters[pos_cluster].get_items_of_cluster():
                        self.solution.get_clusters().append(self.list_clusters.pop(pos_cluster))
                    else:
                        self.list_clusters.pop(pos_cluster)

                    self.list_depots.pop(pos_depot_matrix)
                    total_depots -= 1
                    change = True
        
    def finish(self) -> Solution:       
        pos_element = -1
        i = 0
        while i < len(self.list_clusters):
            clusters = self.list_clusters
            cluster = clusters[i]
            
            print("\n")
            print(f"ID Cluster: {cluster.get_id_cluster()}")
            print(f"Position Cluster: {cluster.get_items_of_cluster()}")
            print(f"Request Cluster: {cluster.get_request_cluster()}")

            pos_element = Problem.get_problem().find_pos_customer(Problem.get_problem().get_customers(), cluster.get_id_cluster())

            if pos_element != -1:
                for item in cluster.get_items_of_cluster():
                    self.solution.get_unassigned_items().append(int(item))
                clusters.remove(cluster)
            else:
                i += 1
         
        if self.list_clusters:
            for cluster in self.list_clusters:
                if cluster.get_items_of_cluster():
                    self.solution.get_clusters().append(cluster) 

        OSRMService.clear_distance_cache()
        
        return self.solution
    
    # Método encargado de obtener el id del depósito con mayor capacidad de la lista.
    def get_id_cluster_with_mu(self, depots: List[Depot], clusters: List[Cluster]) -> int:
        id_depot_mu = depots[0].get_id_depot()
        pos_cluster = self.find_cluster(id_depot_mu, clusters)
        request_cluster = clusters[pos_cluster].get_request_cluster()
        max_capacity_depot = Problem.get_problem().get_total_capacity_by_depot(depots[0])
        max_capacity_depot -= request_cluster
        total_depots = len(depots)

        current_capacity_depot = 0

        for i in range(1, total_depots):
            pos_cluster = self.find_cluster(depots[i].get_id_depot(), clusters)
            request_cluster = clusters[pos_cluster].get_request_cluster()
            current_capacity_depot = Problem.get_problem().get_total_capacity_by_depot(depots[i])
            current_capacity_depot -= request_cluster

            if max_capacity_depot < current_capacity_depot:
                max_capacity_depot = current_capacity_depot
                id_depot_mu = depots[i].get_id_depot()

        return id_depot_mu